prov/efa/src/efa_mr.c

```diff
@@ -176,32 +205,121 @@ struct fi_ops_mr efa_domain_mr_cache_ops = {
 	.regattr = efa_mr_cache_regattr,
 };
 
+static int efa_mr_release(struct efa_mr *mr)
+{
+	struct efa_domain *efa_domain;
+	int ret = 0;
+	int err;
+
+	efa_domain = mr->domain;
+	err = -ibv_dereg_mr(mr->mr);
+	if (!err) {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to close MR\n");
+		ret = err;
+	}
+	err = ofi_mr_map_remove(&efa_domain->util_domain.mr_map,
+				mr->mr_fid.key);
+	if (err && err != -FI_ENOKEY) {
+		EFA_WARN(FI_LOG_MR,
+			"Unable to remove MR entry from util map (%s)\n",
+			fi_strerror(-ret));
+		ret = err;
+	}
+	if (rxr_env.enable_shm_transfer && mr->shm_mr) {
+		err = fi_close(&mr->shm_mr->fid);
+		if (err) {
+			EFA_WARN(FI_LOG_MR,
+				"Unable to close shm MR\n");
+			ret = err;
+		}
+	}
+	return ret;
+}
+
 static int efa_mr_close(fid_t fid)
+
 {
-	struct efa_mem_desc *mr;
+	struct efa_mr *mr;
 	int ret;
 
-	mr = container_of(fid, struct efa_mem_desc, mr_fid.fid);
-	ret = -ibv_dereg_mr(mr->mr);
-	if (!ret)
-		free(mr);
+	mr = container_of(fid, struct efa_mr, mr_fid.fid);
+	ret = efa_mr_release(mr);
```

Comment:
We should check the return value.

Reply:
I think want want to add an FI_WARN here but still call ```free(mr)```.